hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/DataInputNavigableJsonAdapter.java

```diff
@@ -88,34 +90,78 @@ private boolean isQuote() throws IOException {
         return input.readByte() == '"';
     }
 
-    private static class UTF8Reader extends Reader {
+    static class UTF8Reader extends Reader {
 
-        private final DataInput input;
+        private final CharsetDecoder utf8Decoder;
+        private final ByteBuffer inputBuffer;
+        // required to support read() for multibyte characters
+        private boolean hasLeftoverChar;
+        private int leftoverChar;
 
-        UTF8Reader(DataInput input) {
-            this.input = input;
+        UTF8Reader(BufferObjectDataInput input) {
+            byte[] data = obtainBytes(input);
+            inputBuffer = ByteBuffer.wrap(data);
+            inputBuffer.position(input.position());
+            this.utf8Decoder = Bits.UTF_8.newDecoder();
+        }
+
+        // default read() implementation does not handle multibyte chars well
+        @Override
+        public int read() throws IOException {
+            if (hasLeftoverChar) {
+                hasLeftoverChar = false;
+                return leftoverChar;
+            }
+
+            char[] buffer = new char[2];
+            int charsRead = read(buffer, 0, 2);
+            switch (charsRead) {
+                case -1:
+                    return -1;
+                case 2:
+                    leftoverChar = buffer[1];
+                    hasLeftoverChar = true;
+                    return buffer[0];
+                case 1:
+                    return buffer[0];
+                default:
+                    throw new IllegalStateException("Unexpected result from read: " + charsRead);
+            }
         }
 
         @Override
         public int read(char[] cbuf, int off, int len) throws IOException {
             if (off < 0 || (off + len) > cbuf.length) {
                 throw new IndexOutOfBoundsException();
             }
-            int i = 0;
-            try {
-                for (i = 0; i < len; i++) {
-                    byte firstByte = input.readByte();
-                    char c = Bits.readUtf8Char(input, firstByte);
-                    cbuf[off + i] = c;
-                }
-            } catch (EOFException e) {
-                if (i == 0) {
-                    return -1;
+            if (!inputBuffer.hasRemaining()) {
+                return -1;
+            }
+            CharBuffer charbuffer = CharBuffer.wrap(cbuf, off, len);
+            CoderResult result = utf8Decoder.decode(inputBuffer, charbuffer, true);
+            if (result.isError()) {
+                result.throwException();
+            }
+            if (result.isUnderflow()) {
+                // return number of characters read
+                if (!inputBuffer.hasRemaining()) {
+                    utf8Decoder.flush(charbuffer);
```

Comment:
Not sure: can it happen that even if the decode result is underflow, calling `flush()` fails with an overflow? If yes then the reader may never see the final (not flushed) characters.

Reply:
It is possible that `flush()` may return an overflow result. The “decoding operation” guide indicates one should resize the buffer and try flush again, but in this case we have a given buffer and we can neither change the buffer size nor notify the caller somehow... As a side-note, UTF-8 decoder implementation does not maintain any state and its `flush()` implementation does not write anything to the char buffer.